### ISSUE TITLE
PromQL: add "involved samples" for query stats

### ISSUE THREAD

**Issue by Betula-L** [prometheus/prometheus#6668]
## Proposal

Prometheus will record `timings stats` for every query, such as `evalTotalTime `, `resultSortTime` now. It need more stats to evaluate query performance.

**What is we need?**

Expose stats "involved sample" as a query stats, i.e. `currentSamples` https://github.com/prometheus/prometheus/blob/0eb34299dad0ab997bc9d9363a06d6ea6eb45dad/promql/engine.go#L785 

**Why is it important?**

The `timings stats` is an unstable metric, which may be affected by hardware, query queue, etc., especially it's a random number. Sometimes we wanna evaluate whether query was bad or good, using `timings stats` is not a good idea.

However, `currentSamples` is stable and reliable which precisely records how many samples a query travelled. It also reflect how much computation resource consumed, such as memory or cpu.

**Comment by brian-brazil** [comment]
Current samples is a guage, not a counter, so is updated throughout the query. What you'd want is peak samples.

**Comment by Betula-L** [comment]
https://groups.google.com/forum/#!topic/prometheus-users/kZ2DvZnYHUA

**Comment by Harkishen-Singh** [comment]
I am taking this up. Thank you.

**Comment by roidelapluie** [comment]
I think we want: the number of series, the number of samples and the peak number of samples. Thanks.

**Comment by roidelapluie** [comment]
Expectations:

```
$ curl 127.0.0.1:9090/api/v1/query?query=up\&stats=true |python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   354  100   354    0     0   345k      0 --:--:-- --:--:-- --:--:--  345k
{
    "data": {
        "result": [
            {
                "metric": {
                    "__name__": "up",
                    "instance": "127.0.1.1:9090",
                    "job": "prometheus"
                },
                "value": [
                    1582924276.93,
                    "1"
                ]
            }
        ],
        "resultType": "vector",
        "stats": {
            "samples": {
                "peakSamples": 9000,
                "totalSamples": 12000,
                "totalEvaluators": 2,
              },
            "timings": {
                "evalTotalTime": 5.4748e-05,
                "execQueueTime": 6.421e-06,
                "execTotalTime": 6.4869e-05,
                "innerEvalTime": 1.4978e-05,
                "queryPreparationTime": 3.7253e-05,
                "resultSortTime": 0
            }
        }
    },
    "status": "success"
```

**Comment by Harkishen-Singh** [comment]
@roidelapluie not sure about the `evaluatorsCount`, but rest all looks :+1:

**Comment by Harkishen-Singh** [comment]
@brian-brazil any views?

**Comment by bboreham** [comment]
At bug-scrub: we believe this was addressed by #10369.
